hexiwear: added support for additional hardware

davidzuhn opened the pull request:
This PR addresses the issues requested on the previous PR.

Device & pinmux setup for SPI 0-2
Docking station Click interface pin definitions
Additional links to onboard sensor datasheets
Pinmux setup for builtin device GPIO
Sample programs for these newly supported devices

Signed-off-by: david d zuhn <zoo@StateBeltRailway.org>

Commit:
arm: soc: nordic nRF52: Add MPU support

We now have generic ARM M4 MPU support added to Zephyr.
Let's enable it for use with Nordic nRF52 chips.

Memory Layout was generated from Section 8.3 "Memory
Map" of nRF52 Product Specifications (for both nRF52832
and nRF52840):
0x00000000: Flash
0x10000000: Factory Information Config Registers
0x10001000: User Information Config Registers
0x20000000: SRAM
0x40000000: APB Peripherals
0x50000000: AHB Peripherals
0xE0000000: ARM M4 Private Peripheral Registers

NOT Configured:
0x60000000: External RAM
0x80000000: External RAM
0xA0000000: External Device
0xC0000000: External Device

NOTE: More work will be needed for future Nordic MWU (Memory
Watching Unit) support.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
arm: core: mpu: Add core support to NXP MPU

This patch add arm core MPU support to NXP MPU driver.

With this feature it is now possible to enable stack guarding on NXP
MPUs.

Signed-off-by: Vincenzo Frascino <vincenzo.frascino@linaro.org>

fallrisk committed:
arch: arm: Fix SoC issues with Atmel SAM4S series.

Signed-off-by: Justin Watson <jwatson5@gmail.com>

mnkp committed:
watchdog: atmel_sam: enable build for all SAM family

atmel_sam watchdog driver was temporarily limited to SAME70
series only. Now that all SAM series are using ASF the
change can be reverted.

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

galak reviewed (changes requested):
Please break this up into several commits - one per logical change.  Also, please address the issues reported by shippable.

You will also need to rebase this against the latest arm branch, sorry for the annoyance there.

thanks

davidzuhn committed:
hexiwear: add RGB LED and haptic motor configuration

Configure the Pinmux & GPIO lines for the RGB LED.

Optionally configure the haptic feedback motor
  (use CONFIG_HAPTIC_FEEDBACK=y)

Sample application for blinking all three RGB colors
and optionally use the motor.

Signed-off-by: david d zuhn <zoo@StateBeltRailway.org>

davidzuhn committed:
hexiwear: add SPI & docking station support

Configure pinmux for SPI0-2 (based on CONFIG_SPI and CONFIG_SPI_x).

Add definitions for SPI CS, GPIO, and UART lines for the docking
station Click board interfaces.

Sample program to read a rotary encoder and use the LEDs on the
Click Rotary-[RGBYO] boards.

Signed-off-by: david d zuhn <zoo@StateBeltRailway.org>

davidzuhn committed:
hexiwear:  add docking station GPIO demo application

Use the LEDs associated with each Click board slot on
the hexiwear docking station.

Signed-off-by: david d zuhn <zoo@StateBeltRailway.org>

davidzuhn committed:
hexiwear: add demo program for multiple SPI devices

Use the 4x4 button click and rotary-G click to demonstrate
SPI inputs and outputs

Signed-off-by: david d zuhn <zoo@StateBeltRailway.org>

davidzuhn committed:
hexiwear: limit 4x4 demo to hexiwear_k64

Signed-off-by: david d zuhn <zoo@StateBeltRailway.org>

davidzuhn committed:
hexiwear: only define GPIO/PORT when needed

Signed-off-by: david d zuhn <zoo@StateBeltRailway.org>

davidzuhn committed:
hexiwear: style & indentation fixups

Signed-off-by: david d zuhn <zoo@StateBeltRailway.org>

davidzuhn committed:
hexiwear: fix checkpatch warnings on sample programs

Signed-off-by: david d zuhn <zoo@StateBeltRailway.org>

dbkinder commented on samples/boards/hexiwear/README.rst:
```diff
@@ -0,0 +1,23 @@
+Hexiwear Demonstration Applications
+===================================
```
Document titles need to use ###### for the underline (H1)

dbkinder reviewed (changes requested):
one doc tweak -- otherwise looks good

davidzuhn committed:
hexiwear: fix README formatting

Signed-off-by: david d zuhn <zoo@StateBeltRailway.org>

davidzuhn committed:
Merge branch 'arm' into zoo/hexiwear

MaureenHelm commented on samples/boards/hexiwear/rotary-leds/src/main.c:
```diff
@@ -0,0 +1,321 @@
+{
+        spi_dev = device_get_binding(CONFIG_SPI_0_NAME);
+        config.config = SPI_WORD(8);
+        config.max_sys_freq = 256;
```
This is only 256 bps. Is that really what you want?

MaureenHelm commented on boards/arm/hexiwear_k64/doc/hexiwear_k64.rst:
```diff
@@ -100,6 +100,12 @@ The K64F SoC has five pairs of pinmux/gpio controllers.
+| PTA13 | GPIO            | Click 2 I/O button/LED    |
++-------+-----------------+---------------------------+
+| PTA14 | GPIO            | Click 3 I/O button/LED    |
++-------+-----------------+---------------------------+
```
These changes should be in the commit where you added the gpios to the board, not in the commit adding a new demo application

MaureenHelm commented on samples/boards/hexiwear/dock-blinky/README.rst:
```diff
@@ -0,0 +1,35 @@
+
+The three right-most DIP switches (labelled LED1..3) need to be in the UP
+position (LED position).  When these are in the down position, the
+corresponding pushbuttons are usable when the GPIO is set for GPIO_DIR_IN.
```
I'd prefer we modify one of the existing blinky samples (basic/blinky or basic/disco) rather than adding a new sample.

MaureenHelm commented on samples/boards/hexiwear/4x4-buttons/testcase.ini:
```diff
@@ -1,9 +1,5 @@
 [test]
 build_only = true
 tags = samples
+platform_whitelist = hexiwear_k64
```
Squash this commit into the previous commit

MaureenHelm commented on boards/arm/hexiwear_k64/Kconfig.defconfig:
```diff
@@ -68,7 +68,7 @@ config PINMUX_MCUX_PORTD
 
 config PINMUX_MCUX_PORTE
-	def_bool y
+	def_bool n
```
We need this enabled if UART_MCUX_4 (for Bluetooth)

MaureenHelm commented on boards/arm/hexiwear_k64/Kconfig.defconfig:
```diff
@@ -11,10 +11,10 @@ config BOARD
+	bool "Docking Station"
+	default n
+	help
+	  Has Hexiwear docking station
```
Squash these style fixes into their original commits. In this particular case, the commit that introduced config HEXIWEAR_DOCKING_STATION

dbkinder reviewed (approved):
+1 for docs

MaureenHelm commented:
Hi @davidzuhn, thanks again for the PR. Any interest in picking this back up and getting into a merge-able state?

davidzuhn commented:
At this point, I don't have immediate plans to work on this again.   I was working on evaluation of Zephyr, and learning how to make things work with the hexiwear.   And I'm more than willing to contribute back the work I did.   But the amount of work needed to "properly" interact with the Zephyr Project is daunting.   The back and forth on expectations (too many commits, not enough commits, move this to that one, etc) is more than I wish to worry about when what I really want to do is work on a specific application.    The time spent trying to find and deal with issues in the CI system (that I don't see how I can test for before hand) is frustrating.   

Please feel free to do whatever you like with these patches.

nashif commented:
outdated and needs a major rebase, please reopen after fixing the outstanding issues and rebasing it.
